haystack/document_stores/opensearch.py

```diff
@@ -798,6 +1009,136 @@ def _get_raw_similarity_score(self, score):
+                    )
+
+                # see FAISS doc for details: https://github.com/facebookresearch/faiss/wiki/FAQ#can-i-ignore-warning-clustering-xxx-points-to-yyy-centroids
+                min_points_per_cluster = 39
+                n_clusters = 2**code_size
+                if len(documents) < n_clusters:
+                    raise DocumentStoreError(
+                        f"PQ training requires the number of training samples to be greater than or "
+                        f"equal to the number of clusters. Number of provided training samples is `{len(documents)}` "
+                        f"and the number of clusters is `{n_clusters}`."
+                    )
+                if len(documents) < n_clusters * min_points_per_cluster:
+                    logger.warning(
+                        "Consider increasing the number of training samples to at least "
+                        "`%i` to get a reliable PQ index.",
+                        n_clusters * min_points_per_cluster,
+                    )
+
+                encoder = {"name": "pq", "parameters": {"m": m, "code_size": code_size}}
+                settings += f" m:{m} code_size:{code_size}"
+                training_req_body["method"]["parameters"]["encoder"] = encoder
+
+            training_req_body["description"] = settings
+            logger.info("Training IVF index '%s' using {len(documents)} embeddings.", index)
+            train_endpoint = f"/_plugins/_knn/models/{index}-ivf/_train"
+            response = self.client.transport.perform_request(
+                "POST", url=train_endpoint, headers=headers, body=training_req_body
+            )
+            ivf_model = response["model_id"]
+
+            # Wait until model training is finished, _knn_model_trained uses a retry decorator
+            if self._knn_model_trained(ivf_model, headers=headers):
+                logger.info("Training of IVF index '%s' finished.", index)
+
+            # Delete temporary training index
+            self.client.indices.delete(index=f".{index}_ivf_training", headers=headers)
+
+            # Clone original index to temporary one
+            self.client.indices.add_block(index=index, block="read_only")
+            self.client.indices.clone(
+                index=index, target=f".{index}_temp", body={"settings": {"index": {"blocks": {"read_only": False}}}}
+            )
+            self.client.indices.put_settings(index=index, body={"index": {"blocks": {"read_only": False}}})
+            self.client.indices.delete(index=index)
+
+            # Reindex original index to newly created IVF index
+            self._create_document_index(index_name=index, headers=headers)
+            self.client.reindex(body={"source": {"index": f".{index}_temp"}, "dest": {"index": index}})
+            self.client.indices.delete(index=f".{index}_temp")
+
+    @retry(retry=retry_if_not_result(bool), wait=wait_exponential(min=1, max=10))
```

Comment:
I thought about a stop criterion, but it's hard to choose a time window within we expect it to finish. So let's leave it like that.